Milestone 2:  Initialize Redux store and connect given components to state

moinkhanif opened the pull request:
This PL focuses on creating:
- Redux Store
  - `reducers`
  - `actions`
  - `containers`
  - connecting redux store

moinkhanif committed:
Create components and containers folder in src

moinkhanif committed:
Create actions folder with index.js for redux actions

moinkhanif committed:
Add books.js and index.js to new reducers directory

moinkhanif committed:
Implement App component with BookList and BookForm

moinkhanif committed:
Implement BookList with table content

moinkhanif committed:
Implement Book component with default book props

moinkhanif committed:
Implement BooksForm with given book categories

moinkhanif committed:
Define redux actions and reducers

moinkhanif committed:
Initialize Redux Store and connect required component to state

sumancrest0001 commented:
### Status: Required changes :gear:

**Hello @moinkhanif**, You did a good job in implementing requirements from the milestone-2. We are almost done with this milestone. There are minor changes required before moving ahead. Let's have a look at them one by one.

sumancrest0001 commented on src/components/BooksList/BooksList.component.jsx:
```diff
@@ -0,0 +1,36 @@
+    category: PropTypes.string,
+  })),
+};
+
+BookList.defaultProps = {
+  books: {
+    id: 'Undefined',
+    title: 'Undefined',
+    category: 'Undefined',
+  },
+};
+
```
**[Required]** Is it required to pass these default props? Consider a book with an Undefined id, title, and category. I think it doesn't make sense.

sumancrest0001 commented on src/reducers/books.js:
```diff
@@ -0,0 +1,20 @@
+import { CREATE_BOOK, REMOVE_BOOK } from '../actions/index';
+
+const booksReducer = (state = [], action) => {
+  switch (action.type) {
+    case CREATE_BOOK:
+      return {
+        ...state,
+        newbook: action.payload,
+      };
+    case REMOVE_BOOK:
+      return {
+        ...state,
+        REMOVE_BOOK: true,
+      };
+    default:
+      return state;
+  }
+};
```
**[Required]** The bookReducer function is not correct. I think you are mismatching actions and reducer. For example, We get {[existingbookArray, newbook: book]} when CREATE_BOOK is dispatched. In the REMOVE_BOOK you are not passing **id** the book that needs to be deleted. How does it know which book to remove? We can pass an id as a payload and filter the book from bookArray. [Reference](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/react_reducx/bookstore/milestone2/ReducxReducers/CreateAndRemoveReducers.png)

sumancrest0001 commented on src/components/BooksForm/BooksForm.component.jsx:
```diff
@@ -0,0 +1,22 @@
+const BOOK_CATEGORY = ['Action', 'Biography', 'History', 'Horror', 'Kids', 'Learning', 'Sci-Fi'];
```
**[Required]** BooksForm component should be in the container directory.

sumancrest0001 reviewed (changes requested):
**[Optional]** There are a couple of console errors. You can fix them.

**Please make the requested changes and request for another code review. 
If you have any queries regarding this review, message me on Slack
Suman Shrestha**

**Happy coding !!!**

moinkhanif committed:
Fix PrpTypes validation issue with both BookList and Books component

moinkhanif committed:
Move BookForm to containers folder

moinkhanif committed:
Fix redux actions

moinkhanif commented:
Hey @sumancrest0001 , thank you for a detailed review. I have implemented the following changes based on your feedback:

- Set the props to `required` so as to avoid the default props issue.
- Fixed all console errors. Turns out they were all due to incorrect PropTypes.
- Converted `CREATE_BOOK` and `REMOVE_BOOK` into functions that takes in `book` and current state as arguments and return a new State. I also made sure not to mutate the previous state in the process.
- Moved `BooksForm` to the container folder.

Torres-ssf reviewed (approved):
### Milestone Approved :+1: 
Hi @moinkhanif :hand: 
Nice work implementing the requested changes.
Good luck with the next PR.

Happy coding :computer: 
**Sergio Torres**
